src/Config/Converters/DataSourceConverterFactory.cs

```diff
@@ -0,0 +1,136 @@
+    /// value or not while deserializing.</param>
+    internal DataSourceConverterFactory(bool replaceEnvVar)
+    {
+        _replaceEnvVar = replaceEnvVar;
+    }
+
+    private class DataSourceConverter : JsonConverter<DataSource>
+    {
+        // Determines whether to replace environment variable with its
+        // value or not while deserializing.
+        private bool _replaceEnvVar;
+
+        /// <param name="replaceEnvVar">Whether to replace environment variable with its
+        /// value or not while deserializing.</param>
+        public DataSourceConverter(bool replaceEnvVar)
+        {
+            _replaceEnvVar = replaceEnvVar;
+        }
+
+        public override DataSource? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            DataSource dataSource = new(DatabaseType.MSSQL, string.Empty, null);
+            if (reader.TokenType is JsonTokenType.StartObject)
+            {
+
+                while (reader.Read() && reader.TokenType is not JsonTokenType.EndObject)
+                {
+                    if (reader.TokenType is JsonTokenType.PropertyName)
+                    {
+                        string propertyName = reader.GetString() ?? string.Empty;
+                        reader.Read();
+                        switch (propertyName)
+                        {
+                            case "database-type":
+                                dataSource = dataSource with { DatabaseType = EnumExtensions.Deserialize<DatabaseType>(reader.DeserializeString(_replaceEnvVar)!) };
+                                break;
+                            case "connection-string":
+                                dataSource = dataSource with { ConnectionString = reader.DeserializeString(replaceEnvVar: _replaceEnvVar)! };
+                                break;
+                            case "options":
+                                if (reader.TokenType == JsonTokenType.Null)
+                                {
+                                    dataSource = dataSource with { Options = null };
+                                    break;
+                                }
+                                else
+                                {
+                                    Dictionary<string, object?> optionsDict = new();
+                                    while (reader.Read() && reader.TokenType is not JsonTokenType.EndObject)
+                                    {
+                                        string optionsSubproperty = reader.GetString()!;
+                                        reader.Read();
+                                        object? optionsSubpropertyValue;
+                                        if (reader.TokenType is JsonTokenType.String)
+                                        {
+                                            string stringValue = reader.DeserializeString(replaceEnvVar: _replaceEnvVar)!;
```

Comment:
comment right here would be helpful pointing out that this is where you determine whether to resolve the environment variable or keep as-is